src/frequenz/sdk/timeseries/_moving_window.py

```diff
@@ -149,16 +179,35 @@ async def _run_impl(self) -> None:
         try:
             async for sample in self._resampled_data_recv:
                 log.debug("Received new sample: %s", sample)
-                self._buffer.update(sample)
+                if self._resampler and self._avg_sender:
+                    await self._avg_sender.send(sample)
+                else:
+                    self._buffer.update(sample)
+
         except asyncio.CancelledError:
             log.info("MovingWindow task has been cancelled.")
             raise
 
         log.error("Channel has been closed")
 
     async def stop(self) -> None:
-        """Cancel the running task and stop the MovingWindow."""
+        """Cancel the running tasks and stop the MovingWindow."""
         await cancel_and_await(self._update_window_task)
+        if self._resampler_task:
+            await cancel_and_await(self._resampler_task)
+
+    def _configure_resampler(self) -> None:
+        """Configure the components needed to run the resampler."""
+        assert self._resampler is not None
+
+        async def sink_buffer(sample: Sample) -> None:
+            if sample.value is not None:
+                self._buffer.update(sample)
+
+        avg_channel = Broadcast[Sample]("average")
+        self._avg_sender = avg_channel.new_sender()
+        self._resampler.add_timeseries("avg", avg_channel.new_receiver(), sink_buffer)
+        self._resampler_task = asyncio.create_task(self._resampler.resample())
```

Comment:
Looking at this, I'm starting to feel that we should have a resampler that don't work on channels at all, it makes things more complicated (an inefficient) to go through a channel here, where you could have just call `resampler.add(sample)` when handling the data:

```py
            async for sample in self._resampled_data_recv:
                log.debug("Received new sample: %s", sample)
                if self._resampler:
                    self._reampler.add(sample)
                else:
                    self._buffer.update(sample)
```

There is already a `_ResampleHelper` to handle the resampling of only one metric, but it doesn't run its own timer, so we would need to have something that does handle its own timer and ideally can resample multiple metrics.

If you agree this is a good idea, I will turn this into an issue.

Big parenthesis: with historical data, we'll need to build windows without a timer, just assuming data comes ordered and looking at the sample timestamp to trigger the building of a new resampling window, so maybe we don't even need the timer after all.

Reply:
Actually I was looking for a way to avoid using the channel for this particular use case when I started the implementation. I agree on the idea in a separate issue to move faster with this current PR, and then we can get rid of channel here once the issue is implemented. @matthias-wende-frequenz @sahas-subramanian-frequenz do you agree?

Reply:
Agreed!

Reply:
Actually the resampler already support this, because it doesn´t depend on channels but on async iterators. So it is a bit convoluted, but you could have an [async generator object](https://peps.python.org/pep-0525/#asynchronous-generator-object) to send samples to the resampler without creating a channel.

It will still be async, so it's not as efficient as it could be, and also it is convoluted to use, so I think we should still go for a simpler solution.

https://github.com/frequenz-floss/frequenz-sdk-python/issues/303